tests/tests_integration/test_api/test_assets.py

```diff
@@ -245,3 +176,203 @@ def test_filter_by_geo_location(self, cognite_client):
 
         finally:
             cognite_client.assets.delete(id=a.id)
+
+
+def generate_orphan_assets(n_id, n_xid, sample_from):
+    # Orphans only: We link all assets to an existing asset (some by ID, others by XID):
+    s = random_string(20)
+    id_assets = [
+        Asset(name="a", external_id=f"child-by-id-{i}-{s}", parent_id=parent.id)
+        for i, parent in enumerate(random.sample(sample_from, k=n_id))
+    ]
+    xid_assets = [
+        Asset(name="a", external_id=f"child-by-xid-{i}-{s}", parent_external_id=parent.external_id)
+        for i, parent in enumerate(random.sample(sample_from, k=n_xid))
+    ]
+    # Shuffle for good measure ;)
+    random.shuffle(assets := id_assets + xid_assets)
+    return assets
+
+
+def create_asset_tower(n):
+    xid = f"test-tower-{{}}-{random_string(15)}"
+    props = dict(
+        name=random_string(10),
+        description=random_string(10),
+        metadata={random_string(5): random_string(5)},
+        source=random_string(10),
+    )
+    return [
+        Asset(external_id=xid.format(0), **props),
+        *(Asset(external_id=xid.format(i), parent_external_id=xid.format(i - 1), **props) for i in range(1, n)),
+    ]
+
+
+@contextmanager
+def create_hierarchy_with_cleanup(client, assets, upsert=False, upsert_mode=""):
+    # Do not call this auto-cleanup method for asset hierarchies expected to fail, it will
+    # just cause you more misery since delete also will fail in 'finally' block :)
+    assert len(assets) <= 1000, "cleanup might fail if we can't fit all in 1 request"
+    try:
+        yield client.assets.create_hierarchy(assets, upsert=upsert, upsert_mode=upsert_mode)
+    finally:
+        if isinstance(assets, AssetHierarchy):
+            assets = assets._assets
+        client.assets.delete(external_id=[a.external_id for a in assets])
+
+
+@pytest.fixture(scope="class")
+def set_create_lim(cognite_client):
+    with pytest.MonkeyPatch.context() as mp:
+        # We set a low limit to hopefully detect bugs in how resources are split (+threading)
+        # without unnecessarily overloading the API with many thousand assets/request:
+        mp.setattr(cognite_client.assets, "_CREATE_LIMIT", 3)
+        mp.setattr(cognite_client.assets._config, "max_workers", 2)
+        yield
```

Comment:
we alread have the `set_request_limit` and `set_max_workers` ctx managers in tests/utils

Reply:
I agree it would be better to be more specific about which limits it changes.